src/sim/include/batsim.h

```diff
@@ -0,0 +1,29 @@
+/// \file   batsim.h
+/// \author Bernardo Marques
+/// \date   2023-05
+///
+/// \brief  Battery Simulator - simulates a battery charging
+
+#pragma once
+
+#include <stdint.h>
+
+#define BAT_SIM_CALL_RATE_ms (100)      // rate that bat_sim is called
```

Comment:
Couldn't quite understand this call rate usage as I've seen it in the main function being called on charging calculations

Reply:
Battery capacity is current * time and measured in Ah (or mAh for smaller batteries like those on mobile phones).
to be able to calculate the charge current we need to know how much time has passed with charging current that is calculated before.
A battery with Capacity C mAh will be able to provide C mA for one hour, or C/2 mA for 2 hours. Batteries have a maximum discharge current and discharged faster than that will degrade battery life. Usually a very safe discharge current is 0.1C, a current at wich the battery will discharge in 10h. A battery, just like a motor/generator is a dual machine and charging is the inverse process of discharging.

You can charge a Lead Acid battery with any current and the more the current the faster it charges but, there is a limit that is safe and higher than that the battery will heat up, and, the chance for it to cause explosive gases, that will accumulate in the room they are usually housed, and might cause an explosion if there is any electrical spark (contactors and relays can cause electrical sparks while switching) is higher as current increases. The batteries will also heat up until they enter combustion if charging with excess current.
Usually for a 12V Lead Acid battery when charged with 15V the current is lower than the limit (0.1 * Capacity) and battery life is not degraded (IIRC).

So the battery charges with any voltage higher than the actual battery voltage and draws 0.1 * Capacity Amperes of current when with 15V.

This is what I understood after reading a little about batteries on the internet. Lithium batteries (Li-ion or LiFePo) have a more complicated charge procedure that I do not yet know about and are more prone to overheat and enter combustion but they usually also have more protections built into the battery modules so not to over discharge, overcharge, overheat, they'll disconnect the cells internally, as opposed to Lead Acid batteries that do not have any control integrated in the modules and will be permanently damaged if over-discharged, degraded if overcharged or overheat and will enter combustion because the is no protection built in the battery module.

time is the call rate of the function, the current we calculated from voltage (I'm not certain if the equation I wrote even simulates a battery charging, it's just something I made up, or remember reading about). The charge added to the battery is current * function call rate (disregarding battery losses, supposing that a battery is 100% efficient when charging).

Reply:
in main.c what is calculated is accumulated energy.
Electrical energy is measured in power over time (Wh).
batsim_run() returns power consumed by the battery in Watt and it needs to be muliplied by elapsed time in hours to have Wh.
The amount of hours that have passed is the rate that we call run divided by hours. One hour is 3600 seconds, since call rate is in ms one hour is 3.600.000 call rate. The value for acc_power is in fixed point for mWh, or:
power (W) * call rate (ms) / 3.600.000 (ms/h) * 1000 (fixed point milli)